### DISCUSSION TITLE
CMS and non-CMS Endpoint Business Logic

### DISCUSSION THREAD

**Discussion by Thermal-Shock** [directus/directus#11941] [Archive]
Hopefully I can explain this correctly being a noob to API/CMS servers here. 

We currently have an old-school LAMP stack and we are only just starting to want to break off our own API servers based upon our current MySQL database model for our current non-responsive eCommerce front and admin sites, and other internal operational sites. It looks like Directus is an amazing fit and with custom interfaces Directus admin could even replace our current admin backend. 

Firstly, we have lots of business logic that we would also like to be accessed by an API. They would use the underlying MySQL database, but wouldn't necessarily have a CMS interface. For example, /api/shipping/rate?... etc.

I'm assuming this kind of thing is possible and perhaps pretty common, but I thought I'd make sure. If so, is there an example/short video anywhere that describes creating an endpoint in Directus for business logic like this? 

Secondly, when someone performs an post with the API or uses the CMS admin to create a row, I'm assuming we can also add validation and perform arbitrary code that uses the SQL database or reject the addition based on some business reason?

Thank you so much. I'm pretty excited about using this platform. I thought Strapi was great until I found this project. 

All the best.

**Comment by rijkvanzanten** [comment]
Heya! Both those things are possible! You can create Custom Endpoints to.. create custom endpoints, and use hooks to run any arbitrary business logic based on events that are triggered in the system (like item creations). See https://docs.directus.io/extensions/creating-extensions/ for more information on how to start creating extensions in general, and https://docs.directus.io/extensions/endpoints/ and https://docs.directus.io/extensions/hooks/ to learn more about custom endpoints and hooks respectively.

To do you one better: we're currently working on a no-code automation engine that will allow you to setup a lot of these automation flows without having to write any JavaScript. More news on that hopefully very soon 🙂

**Reply by Thermal-Shock** [reply]
Thank you for the speedy answer and the references Rijk. 

The big effort for us is porting all our PHP based business logic into well organized endpoints, but fortunately we have been writing more and more advanced API based Javascript lately, so I'm hoping the transition will be a fairly straightforward systematic porting exercise. It's critical that we choose and invest in the best API platform we can that also meets the CMS needs of our internal and public sub-sites such as machine assembly instructions and support documentation. 

This automation engine sounds great. Anything that improves the developer experience and quality is awesome. I'll look forward to that. Is there anything I can get subscribed to that will let me know when things like this are published?

If we go self-hosted, I'm assuming it's still pretty straightforward to upgrade? We are a team of strong programmers, but best to ask. Do you anticipate any breaking changes as the platform evolves version to version or is the platform pretty well decided at this point?

Thanks again for the advice. All the best.

**Reply by rijkvanzanten** [reply]
> Is there anything I can get subscribed to that will let me know when things like this are published?

You can "Watch" the repo in the top right of GitHub, and be notified of every release that's being created. That way, you'll be up to speed with all the latest and greatest that happens on the project 🙂 

> If we go self-hosted, I'm assuming it's still pretty straightforward to upgrade? We are a team of strong programmers, but best to ask. Do you anticipate any breaking changes as the platform evolves version to version or is the platform pretty well decided at this point?

The upgrade process should be pretty straightforward, yeah. As for breaking changes, we don't expect any anytime soon. If there would be one, it would be a major version changes as well (eg Directus 9 ➡️ 10)
